File: backend/mcp-code-intelligence-nodejs/src/memory/handoff.ts
/**
 * AgentHandoffMemory — preserves context between agent sessions.
 * Structured handoff records with decisions, questions, artifacts.
 */

export interface HandoffContext {
  fromAgent: string;
  toAgent: string;
  summary: string;
  keyDecisions: string[];
  openQuestions: string[];
  artifacts: string[];
  ticketKey?: string;
}

export class AgentHandoffMemory {
  private repo: any;
  private searchRepo: any;

  constructor(repo: any, searchRepo: any) {
    this.repo = repo;
    this.searchRepo = searchRepo;
  }

  /** Record a handoff between agents. */
  recordHandoff(ctx: HandoffContext): number {
    const content = this.formatHandoff(ctx);
    return this.repo.insert({
      content,
      summary: `Handoff: ${ctx.fromAgent} → ${ctx.toAgent}: ${ctx.summary.slice(0, 60)}`,
      type: 'CONTEXT',
      tier: 'WORKING',
      source: ctx.ticketKey ?? null,
      tags: `handoff,${ctx.fromAgent},${ctx.toAgent}`,
    });
  }

  /** Get recent handoffs for an agent. */
  getHandoffsForAgent(agentName: string, limit = 5): any[] {
    const results = this.searchRepo.searchByTags(['handoff', agentName], limit);
    return results.sort((a: any, b: any) =>
      (b.createdAt ?? '').localeCompare(a.createdAt ?? '')
    );
  }

  /** Get latest handoff context for a ticket. */
  getLatestForTicket(ticketKey: string): any | null {
    const results = this.searchRepo.search(`${ticketKey} handoff`, 1);
    return results[0]?.entry ?? null;
  }

  private formatHandoff(ctx: HandoffContext): string {
    const lines: string[] = [];
    lines.push(`## Agent Handoff: ${ctx.fromAgent} → ${ctx.toAgent}`);
    lines.push(`\n### Summary\n${ctx.summary}`);
    if (ctx.keyDecisions.length > 0) {
      lines.push('\n### Key Decisions');
      ctx.keyDecisions.forEach(d => lines.push(`- ${d}`));
    }
    if (ctx.openQuestions.length > 0) {
      lines.push('\n### Open Questions');
      ctx.openQuestions.forEach(q => lines.push(`- ${q}`));
    }
    if (ctx.artifacts.length > 0) {
      lines.push('\n### Artifacts');
      ctx.artifacts.forEach(a => lines.push(`- ${a}`));
    }
    return lines.join('\n');
  }
}

File: backend/mcp-code-intelligence-python/src/mcp_code_intel/memory/handoff.py
"""AgentHandoffMemory — preserves context between agent sessions."""

from __future__ import annotations

from dataclasses import dataclass, field
from typing import Any


@dataclass
class HandoffContext:
    """Handoff context for agent transitions."""

    from_agent: str
    to_agent: str
    summary: str
    key_decisions: list[str] = field(default_factory=list)
    open_questions: list[str] = field(default_factory=list)
    artifacts: list[str] = field(default_factory=list)
    ticket_key: str | None = None


class AgentHandoffMemory:
    """Records and retrieves agent handoff context."""

    def __init__(self, repo: Any, search_repo: Any) -> None:
        self._repo = repo
        self._search = search_repo

    def record_handoff(self, ctx: HandoffContext) -> int:
        """Record a handoff between agents."""
        content = self._format_handoff(ctx)
        return self._repo.insert_entry(
            content=content,
            summary=f"Handoff: {ctx.from_agent} → {ctx.to_agent}: {ctx.summary[:60]}",
            entry_type="CONTEXT",
            tier="WORKING",
            source=ctx.ticket_key,
            tags=f"handoff,{ctx.from_agent},{ctx.to_agent}",
        )

    def get_handoffs_for_agent(self, agent_name: str, limit: int = 5) -> list[dict]:
        """Get recent handoffs for an agent."""
        results = self._search.search_by_tags(["handoff", agent_name], limit)
        return sorted(results, key=lambda r: r.get("created_at", ""), reverse=True)

    def get_latest_for_ticket(self, ticket_key: str) -> dict | None:
        """Get latest handoff context for a ticket."""
        results = self._search.search(f"{ticket_key} handoff", limit=1)
        return results[0].get("entry") if results else None

    def _format_handoff(self, ctx: HandoffContext) -> str:
        lines: list[str] = []
        lines.append(f"## Agent Handoff: {ctx.from_agent} → {ctx.to_agent}")
        lines.append(f"\n### Summary\n{ctx.summary}")
        if ctx.key_decisions:
            lines.append("\n### Key Decisions")
            lines.extend(f"- {d}" for d in ctx.key_decisions)
        if ctx.open_questions:
            lines.append("\n### Open Questions")
            lines.extend(f"- {q}" for q in ctx.open_questions)
        if ctx.artifacts:
            lines.append("\n### Artifacts")
            lines.extend(f"- {a}" for a in ctx.artifacts)
        return "\n".join(lines)
